File: 0042-trapping-rain-water/0042-trapping-rain-water.java
class Solution {
    public int trap(int[] height) {
        int leftMax = 0;
        int rightMax = 0;

        int left = 0;
        int right = height.length - 1;
        int water = 0;
        while (left < right) {
            if (height[left] < height[right]) {
                leftMax = Math.max(leftMax, height[left]);
                water += leftMax - height[left];
                left++;
            } else {
                rightMax = Math.max(rightMax, height[right]);
                water += rightMax - height[right];
                right--;                
            }
        }

        return water;
    }
}

File: 0042-trapping-rain-water/0042-trapping-rain-water.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int left = 0;
        int right = height.size() - 1;
        int result = 0;

        int leftMax = 0;
        int rightMax = 0;
        while (left < right) {
            if (height[left] < height[right]) {
                leftMax = max(height[left], leftMax);
                result += leftMax - height[left];
                left++;
            } else {
                rightMax = max(height[right], rightMax);
                result += rightMax - height[right];
                right--;                
            }
        }

        return result;
    }
};